Source/Core/Core/AchievementManager.cpp

```diff
@@ -270,6 +276,124 @@ void AchievementManager::ActivateDeactivateRichPresence()
       nullptr, 0);
 }
 
+void AchievementManager::FetchBadges()
+{
+  if (!m_is_runtime_initialized || !IsLoggedIn() || !Config::Get(Config::RA_BADGES_ENABLED))
+  {
+    if (m_update_callback)
+      m_update_callback();
+    return;
+  }
+  m_image_queue.Cancel();
+  if (!m_player_badge.loaded)
+  {
+    m_image_queue.EmplaceItem([this] {
+      {
+        std::lock_guard lg{m_lock};
+        // Ensure badge still needs loading by the time this reaches the top of the queue
+        if (m_player_badge.loaded)
+          return;
+        // m_display_name is safe as only VerifyCredentials modifies it and both calls to
+        // VerifyCredentials are locked
+        rc_api_fetch_image_request_t icon_request = {.image_name = m_display_name.c_str(),
+                                                     .image_type = RC_IMAGE_TYPE_USER};
+        if (RequestImage(icon_request, &m_player_badge) == ResponseType::SUCCESS)
+          INFO_LOG_FMT(ACHIEVEMENTS, "Successfully downloaded player badge id {}.", m_display_name);
+        else
+          WARN_LOG_FMT(ACHIEVEMENTS, "Failed to download player badge id {}.", m_display_name);
+      }
+      if (m_update_callback)
+        m_update_callback();
+    });
+  }
+
+  if (!IsGameLoaded())
+  {
+    if (m_update_callback)
+      m_update_callback();
+    return;
+  }
+
+  if (!m_game_badge.loaded)
+  {
+    m_image_queue.EmplaceItem([this] {
+      {
+        std::lock_guard lg{m_lock};
+        if (m_game_badge.loaded)
+          return;
+        // Only FetchGameData modifies m_game_data and that locks it down while modifying.
+        rc_api_fetch_image_request_t icon_request = {.image_name = m_game_data.image_name,
+                                                     .image_type = RC_IMAGE_TYPE_GAME};
+        if (RequestImage(icon_request, &m_game_badge) == ResponseType::SUCCESS)
+        {
+          INFO_LOG_FMT(ACHIEVEMENTS, "Successfully downloaded game badge id {}.",
+                       m_game_data.image_name);
+        }
+        else
+        {
+          WARN_LOG_FMT(ACHIEVEMENTS, "Failed to download game badge id {}.",
+                       m_game_data.image_name);
+        }
+      }
+      if (m_update_callback)
+        m_update_callback();
+    });
+  }
+
+  unsigned num_achievements = m_game_data.num_achievements;
+  for (size_t index = 0; index < num_achievements; index++)
+  {
+    std::lock_guard lg{m_lock};
+    // In case the number of achievements changes since the loop started; I just don't want
+    // to lock for the ENTIRE loop so instead I reclaim the lock each cycle
+    if (index >= m_game_data.num_achievements)
+      break;
+    const UnlockStatus& unlock_status = m_unlock_map[m_game_data.achievements[index].id];
```

Comment:
This should use `m_unlock_map.find()` unless you intend the possibility of inserting new items here. This also goes for the two accesses in the queue items.

Reply:
Ahh, I always forget that and I really wish it didn't do that.

Reply:
Yeah, the C++ map interface is awful.